File: scripts/UpdateStatusAt/updateStatusAt.js
/**
 * Altera um status de título (UpdateStatusAT) no webservice Acompanhamento de Títulos da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { envInt, envStr, requireEnvInt, resolvePath } from "../../lib/onr_env.js";
import {
  hashErrorHint,
  loadAcompanhamentoSoapConfig,
  loadLoginConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_acompanhamento.js";
import { buildUpdateStatusRequest } from "../../lib/onr_update_status_at.js";

const PREFIX = "ACOMPANHAMENTO_TITULOS_UPDATE_STATUS_";

function resolveIdStatus() {
  let idStatus = envInt(`${PREFIX}ID_STATUS`);
  if (idStatus === undefined) {
    idStatus = envInt("ACOMPANHAMENTO_TITULOS_ID_STATUS");
  }
  if (idStatus === undefined) {
    throw new Error(
      `Defina ${PREFIX}ID_STATUS ou ACOMPANHAMENTO_TITULOS_ID_STATUS no .env.`
    );
  }
  return idStatus;
}

function loadConfig() {
  const required = [
    `${PREFIX}ID_TIPO_STATUS`,
    `${PREFIX}DATA_STATUS`,
    `${PREFIX}DESCRICAO_STATUS`,
  ];
  const missing = required.filter((k) => !process.env[k]);
  if (missing.length) {
    throw new Error(`Variáveis ausentes no .env: ${missing.join(", ")}`);
  }

  const idTipoStatus = requireEnvInt(`${PREFIX}ID_TIPO_STATUS`);
  const dataStatus = envStr(`${PREFIX}DATA_STATUS`);
  const descricaoStatus = envStr(`${PREFIX}DESCRICAO_STATUS`);

  if (!dataStatus?.trim()) {
    throw new Error(
      `Defina ${PREFIX}DATA_STATUS no .env (formato: aaaa-mm-dd hh:mm:ss).`
    );
  }
  if (!descricaoStatus?.trim()) {
    throw new Error(
      `Defina ${PREFIX}DESCRICAO_STATUS no .env (obrigatório na spec — erro 17 se vazio).`
    );
  }

  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    idStatus: resolveIdStatus(),
    idTipoStatus,
    dataStatus: dataStatus.trim(),
    descricaoStatus: descricaoStatus.trim(),
    ...loadAcompanhamentoSoapConfig(),
  };
}

async function updateStatusAt(cfg, oRequest) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });

  let response;
  try {
    [response] = await client.UpdateStatusATAsync({ oRequest });
  } catch (err) {
    const status = err?.response?.status;
    const body =
      typeof err?.body === "string"
        ? err.body
        : typeof err?.response?.data === "string"
          ? err.response.data
          : null;

    if (status === 503 || (body && !body.trimStart().startsWith("<"))) {
      throw new Error(
        `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`
      );
    }
    throw err;
  }

  return (
    response?.UpdateStatusATResult ??
    response?.UpdateStatusATResponse?.UpdateStatusATResult ??
    response
  );
}

async function main() {
  const cfg = loadConfig();
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg);
  const oRequest = buildUpdateStatusRequest(hash, cfg);

  console.log("=== Parâmetros UpdateStatusAT ===");
  console.log(JSON.stringify(oRequest, null, 2));
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await updateStatusAt(cfg, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nUpdateStatusAT falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    if (response.CODIGOERRO === 12) {
      console.error(
        `Dica: status inválido — confira ${PREFIX}ID_STATUS / ACOMPANHAMENTO_TITULOS_ID_STATUS.`
      );
    }
    if (response.CODIGOERRO === 13) {
      console.error(`Dica: ${PREFIX}ID_TIPO_STATUS inválido — veja § 3.2.1.`);
    }
    if ([14, 15, 16].includes(response.CODIGOERRO)) {
      console.error(`Dica: revise ${PREFIX}DATA_STATUS (19 caracteres).`);
    }
    if (response.CODIGOERRO === 17) {
      console.error(`Dica: informe ${PREFIX}DESCRICAO_STATUS.`);
    }
    if (response.CODIGOERRO === 30) {
      console.error("Dica: não foi possível localizar o status (30).");
    }
    if (response.CODIGOERRO === 32) {
      console.error("Dica: sem permissão para alterar o status (spec cód. 32).");
    }
    if (
      response.CODIGOERRO === 0 &&
      response.ERRODESCRICAO?.includes("IDMsg")
    ) {
      console.error(
        "Dica: envie DataStatus e DescricaoStatus no XML; se persistir, informe IDMsg à ONR."
      );
    }
    const hint = hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  console.log(
    `\nOK — Status ${cfg.idStatus} atualizado (tipo ${cfg.idTipoStatus}).`
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: scripts/UpdateStatusAt/updateStatusAt.py
#!/usr/bin/env python3
"""Altera um status de título (UpdateStatusAT) no webservice Acompanhamento de Títulos da ONR."""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_acompanhamento import (  # noqa: E402
    hash_error_hint,
    load_acompanhamento_soap_config,
    load_login_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_env import env_int, env_str, require_env_int  # noqa: E402
from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_update_status_at import build_update_status_request  # noqa: E402
from lib.onr_zeep_serialize import serialize_result  # noqa: E402

_PREFIX = "ACOMPANHAMENTO_TITULOS_UPDATE_STATUS_"


def _resolve_id_status() -> int:
    id_status = env_int(f"{_PREFIX}ID_STATUS")
    if id_status is None:
        id_status = env_int("ACOMPANHAMENTO_TITULOS_ID_STATUS")
    if id_status is None:
        raise SystemExit(
            f"Defina {_PREFIX}ID_STATUS ou ACOMPANHAMENTO_TITULOS_ID_STATUS no .env."
        )
    return id_status


def load_config() -> dict:
    load_dotenv(ROOT / ".env")

    required = [
        f"{_PREFIX}ID_TIPO_STATUS",
        f"{_PREFIX}DATA_STATUS",
        f"{_PREFIX}DESCRICAO_STATUS",
    ]
    missing = [k for k in required if not os.getenv(k)]
    if missing:
        raise SystemExit(f"Variáveis ausentes no .env: {', '.join(missing)}")

    id_tipo_status = require_env_int(f"{_PREFIX}ID_TIPO_STATUS")
    data_status = env_str(f"{_PREFIX}DATA_STATUS")
    descricao_status = env_str(f"{_PREFIX}DESCRICAO_STATUS")

    if not data_status or not data_status.strip():
        raise SystemExit(f"Defina {_PREFIX}DATA_STATUS no .env (formato: aaaa-mm-dd hh:mm:ss).")
    if not descricao_status or not descricao_status.strip():
        raise SystemExit(
            f"Defina {_PREFIX}DESCRICAO_STATUS no .env (obrigatório na spec — erro 17 se vazio)."
        )

    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        "id_status": _resolve_id_status(),
        "id_tipo_status": id_tipo_status,
        "data_status": data_status.strip(),
        "descricao_status": descricao_status.strip(),
        **load_acompanhamento_soap_config(),
    }


def main() -> int:
    cfg = load_config()
    hash_value = resolve_auth_hash(cfg["chave"], cfg["login_cfg"])
    o_request = build_update_status_request(hash_value, cfg)

    print("=== Parâmetros UpdateStatusAT ===")
    print(json.dumps(o_request, ensure_ascii=False, indent=2))
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "UpdateStatusAT", o_request)
    response = to_json_safe(serialize_result(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(f"\nUpdateStatusAT falhou: [{codigo}] {descricao}", file=sys.stderr)
        if codigo == 12:
            print(
                f"Dica: status inválido — confira {_PREFIX}ID_STATUS / "
                "ACOMPANHAMENTO_TITULOS_ID_STATUS (GetStatusAT / InsertStatusAT).",
                file=sys.stderr,
            )
        if codigo == 13:
            print(
                f"Dica: {_PREFIX}ID_TIPO_STATUS inválido — veja tipos em § 3.2.1.",
                file=sys.stderr,
            )
        if codigo in (14, 15, 16):
            print(
                f"Dica: revise {_PREFIX}DATA_STATUS (19 caracteres, ex.: 2026-05-19 10:00:00).",
                file=sys.stderr,
            )
        if codigo == 17:
            print(
                f"Dica: informe {_PREFIX}DESCRICAO_STATUS.",
                file=sys.stderr,
            )
        if codigo == 30:
            print(
                "Dica: não foi possível localizar o status (30). Confira o IDStatus.",
                file=sys.stderr,
            )
        if codigo == 32:
            print(
                "Dica: sem permissão para alterar o status (spec cód. 32).",
                file=sys.stderr,
            )
        if codigo == 0 and descricao and "IDMsg" in descricao:
            print(
                "Dica: envie DataStatus e DescricaoStatus no XML. "
                "Os scripts já incluem ambos; se persistir, informe o IDMsg à ONR.",
                file=sys.stderr,
            )
        hint = hash_error_hint(codigo) if isinstance(codigo, int) else None
        if hint:
            print(hint, file=sys.stderr)
        return 1

    print(
        f"\nOK — Status {cfg['id_status']} atualizado (tipo {cfg['id_tipo_status']})."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
